### PR TITLE
Add TA-131639

### PR DESCRIPTION
**Pull request by rmloveland**
Fixes DOC-11251

Preview: https://deploy-preview-18981--cockroachdb-docs.netlify.app/docs/advisories/a131639.html

### CODE REVIEW

**Review by arulajmani — APPROVED** [review]

**Comment by rmloveland** [comment]
TFTR @arulajmani !

(FYI for anyone reading, the reason this build is failing is b/c the link to v23.1.27 release notes does not exist yet, once https://github.com/cockroachdb/docs/pull/18978 merges that link will exist and I'll trigger re-build of this)

**Comment by nvanbenschoten** [comment]
@rmloveland please hold off on merging this for now. We need to find a better way to eliminate false positives without manual inspection of the logs before publishing. If we don’t do a good enough job, we’ll create a lot of undue concern.

**Comment by rmloveland** [comment]
> @rmloveland please hold off on merging this for now. We need to find a better way to eliminate false positives without manual inspection of the logs before publishing. If we don’t do a good enough job, we’ll create a lot of undue concern.

👍

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:4]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:15]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:19]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:21]: ✅

Comment (rmloveland) [reply src/current/advisories/a131639.md:21]: thx ✅ 

 also checked one other use of this phrase and updated it to capitalize its use of 'Raft'

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:31]: ya let's do it, i was gonna say fixed is clearer but upon reflection addressed is probably the proper term in the sense that a patch addresses an issue but whether it fixes the issue is something empirical that can only be determined over time

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:37]: nice totes agreed, "the patch" fits with the "addresses" language better

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:47]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:71]: > Maybe we can "bold" the bits of interesting (does markdown support bold within a code block??).

AFAICT the answer is no but we can do ASCII gymnastics like

```
INFO: the value of foo=bar at time t=123
                                ^^^^^^
                           LOOK FOR THIS
```

File: src/current/advisories/a131639.md
Comment (florence-crl) [inline src/current/advisories/a131639.md:21]: s/GC-ed/garbage collected/

File: src/current/advisories/a131639.md
Comment (florence-crl) [inline src/current/advisories/a131639.md:73]: s/indices/indexes/
I know that "indices" is also correct, but it is not used the docs. "Indexes" is used throughout the docs instead.

File: src/current/advisories/a131639.md
Comment (florence-crl) [inline src/current/advisories/a131639.md:35]:
```suggestion
[#123442](https://github.com/cockroachdb/cockroach/pull/123442) resolves an issue in CockroachDB in the expiration-to-epoch lease promotion transition, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.
```

**Review by florence-crl — COMMENTED** [review]
Please consider my suggestions. thanks!

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:21]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:73]: ✅

File: src/current/advisories/a131639.md
Comment (rmloveland) [reply src/current/advisories/a131639.md:35]: ✅

**Comment by rmloveland** [comment]
> Please consider my suggestions. thanks!

thanks @florence-crl! applied all of your suggestions

File: src/current/advisories/a131639.md
Comment (nicktrav) [inline src/current/advisories/a131639.md:49]: We now have a [script](https://github.com/cockroachlabs/support/pull/3101) that can be used, given a `debug.zip` as input.

The script requires two dependencies - `ripgrep` and `duckdb`. The install instructions for each can be found here:
- https://github.com/BurntSushi/ripgrep#installation
- https://duckdb.org/docs/installation

@rmloveland - thoughts on how to best include this? We could include the script inline, perhaps at the bottom of the TA, so as not to clutter things to much? Alternatively, we could host it somewhere, and reference it. Will defer to you.

Comment (rmloveland) [reply src/current/advisories/a131639.md:49]: @nicktrav (and @nvanbenschoten as script author) I took a shot at including the script inline, and updated the section with an edited version of the description of https://github.com/cockroachlabs/support/pull/3101

PTAL

one thing i noticed is that the script outputs range IDs and the next section 'Detection via an audit of primary and secondary indexes' refers to checking indexes by name

is there guidance we can offer to users on how to get from the potentially affected range IDs to the names of indexes they need to check? 

AFAICT it's pretty easy to get from a schema object to a range ID with [`SHOW RANGES`](https://www.cockroachlabs.com/docs/v24.2/show-ranges.html) but i'm not sure if we have a supported (or unsupported) way to map from range IDs up to schema objects?  if we don't write it here, it seems folks will reach out to ask

Comment (rmloveland) [reply src/current/advisories/a131639.md:49]: found some discussion [here](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1716323894389819) of how to do approx a range ID -> SQL mapping and it looks pretty "internal" / unsupported / confusing but i defer to yall

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [inline src/current/advisories/a131639.md:35]: @rmloveland do we prefer to point to Pull Requests from TAs, or to commits (in this case, https://github.com/cockroachdb/cockroach/commit/6dd54b46cc56b7d2b302e0d5ec1509658a1c86f7)? The pull request includes the full code review, which is not relevant for a TA.

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [inline src/current/advisories/a131639.md:65]: This output is a little stale and also includes more information than we should probably share externally. How about:
```bash
➜ pwd
/path/of/unzipped/debug

➜ /path/to/script/ta131639_check.sh
Searching for liveness epoch increments ...
Searching for slow lease promotions ...
Querying the logs for lease expiration regressions ...

❌ Lease expiration regressions found. Symptoms detected.

symptoms
node 123 observed possible symptoms at 2024-09-01 12:34:56 on ranges 123456
```

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [inline src/current/advisories/a131639.md:51]: The script to longer accepts a debug.zip. Instead, it must be run from within an unzipped debug.zip directory.

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [inline src/current/advisories/a131639.md:185]: Should we be more prescriptive about why someone might suspect that they were subject to a lost write? Something like:
"If the detection script detects symptoms of a potential lease expiration regression and you suspect that a particular row may have been subject to a lost write ... "

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [reply src/current/advisories/a131639.md:49]: This is a good question. Shaun M closes that thread with "The fact that a range can contain primary indexes, secondary indexes, and even have multiple tables within the same range - there’s a many-to-one relationship which is rough to untangle". That's pretty much right, and so it's not as clear an answer as it might seem to perform this mapping, at least not in an automated way. Sean C includes the kind of query we'd need to run to perform the mapping, but it's involved.

If we get to this point with a customer where the script detecting symptoms is pointing at range IDs, it might just be best to get support involved. The next steps are going to be different on a case-by-case basis.

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:51]:
```suggestion
Users can run the script [a131639_check.sh](a131639_check.sh) to determine whether any lease expiration regressions are evident from their logs. The script must be run from within a decompressed [debug.zip]({% link v24.3/cockroach-debug-zip.md %}) directory.
```

@nvanbenschoten My suggestion here (1) links to a copy of the script in the docs repo (2) implements your comment (3) removes the language that I found too vague that the script will "detect part of the symptoms of this technical advisory".

I presume users would benefit from a more specific caveat about what the script may or may not find, but LMK what you think, or if this is OK as-is.

File: src/current/advisories/a131639.md
Comment (mikeCRL) [reply src/current/advisories/a131639.md:49]: Thanks, @nvanbenschoten. Given the various callouts we have to contact Support, can we leave the copy as-is, or is there anything you'd add to point them in the right direction about what to expect, even if the end result is needing to contact Support?

@nickelnickel can you also review this, please?

**Review by mikeCRL — COMMENTED** [review]
@nvanbenschoten Thanks. Back to you.

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:7]:
```suggestion
advisory_date: 2024-10-07
```

File: src/current/advisories/a131639.md
Comment (mikeCRL) [reply src/current/advisories/a131639.md:35]:
```suggestion
[In #123442](https://github.com/cockroachdb/cockroach/commit/6dd54b46cc56b7d2b302e0d5ec1509658a1c86f7), we resolved an issue with CockroachDB in the expiration-to-epoch lease promotion transition process, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.
```

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:181]: Suggesting that we drop the code, in lieu of the suggested link above

```suggestion
```

File: src/current/advisories/a131639.md
Comment (mikeCRL) [reply src/current/advisories/a131639.md:185]: @nvanbenschoten This SGTM. Will implement after your re-review. (Sorry, I missed some of your comments as we were both working with the PR at the same time; I think I've now addressed all of them.)

File: src/current/advisories/a131639.md
Comment (mikeCRL) [reply src/current/advisories/a131639.md:49]: @nvanbenschoten: Nick (@nickelnickel) tells me he's good with this as-is, so if the copy's good to you, we are good to go.

Comment (nvanbenschoten) [reply src/current/advisories/a131639.md:49]: I'm good with this as is.

File: src/current/advisories/a131639.md
Comment (nvanbenschoten) [reply src/current/advisories/a131639.md:35]: This suggestion sounds good to me.

File: src/current/advisories/a131639_check.sh
Comment (nvanbenschoten) [inline src/current/advisories/a131639_check.sh:2]: @mikeCRL `s/unziped/unzipped/` (my typo, but mind fixing here?)

File: src/current/advisories/a131639.md
Comment (mdlinville) [inline src/current/advisories/a131639.md:15]:
```suggestion
In v22.2 through v24.2, in circumstances where there is a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddle multiple ranges to be lost. In cases where these lost writes are not also present on one or more secondary indexes that could allow for reconstructing or repairing the row, these writes could be irrecoverably lost.
```
Slight edit here because I think the "Since...through..." construct reads a little confusing. I'm confused about the distinction being made between "lost" and "irrevocably lost", and I suspect in the next paragraph, the first and second sentences are saying the same thing.

File: src/current/advisories/a131639.md
Comment (mdlinville) [reply src/current/advisories/a131639.md:29]: Would it be better to list impacted  v22.2 versions also, since that's the only other version that is impacted? So then it would be: v22.2.0-19, v23.1.0-26, v23.2.0-10, and v24.1.0.

**Comment by rytaft** [comment]
@rmloveland I made some suggested changes to the doc here: https://docs.google.com/document/d/1QNqNR-hfnEUD8K4wlBiOSqgF8W79zlXTP7WoDhWFAkU/edit?usp=sharing. Any chance you can pull those changes into this PR? Thank you!

File: src/current/advisories/a131639_check.sh
Comment (mikeCRL) [reply src/current/advisories/a131639_check.sh:2]: 👍 
```suggestion
# Run from within an unzipped debug.zip directory.
```

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:6]:
```suggestion
affected_versions: v22.2, v23.1.0 to v23.1.26, v23.2.0 to v23.2.10, v24.1.0
```

adding 22.2 major version to match another recent TA

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:27]:
```suggestion
- v22.2
- v23.1.0 - v23.1.26
```

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:57]:
```suggestion
The script requires the following dependencies:
```

File: src/current/advisories/a131639.md
Comment (nicktrav) [reply src/current/advisories/a131639.md:27]: LGTM

File: src/current/advisories/a131639.md
Comment (mikeCRL) [inline src/current/advisories/a131639.md:25]:
```suggestion
The following CockroachDB versions are impacted by this bug:
```

**Pull request merged**
